File: rnef-backtest/backend/metrics.py
import numpy as np
import pandas as pd


def apply_tx_cost(weights_before: np.ndarray, weights_after: np.ndarray, tx_cost_bps: int) -> float:
    turnover = np.sum(np.abs(weights_after - weights_before))
    return turnover * (tx_cost_bps / 10_000)


def compute_metrics(weights: np.ndarray, returns_df: pd.DataFrame, tx_cost_bps: int):
    mean_ret = returns_df.mean().values
    cov = returns_df.cov().values
    equal = np.ones(len(weights)) / len(weights)

    port_return = float(np.dot(weights, mean_ret) * 252)
    port_vol = float(np.sqrt(weights @ cov @ weights * 252))
    sharpe = port_return / port_vol if port_vol > 0 else 0.0

    cost = apply_tx_cost(equal, weights, tx_cost_bps)
    port_return_net = port_return - cost * 252

    return round(port_return_net, 4), round(port_vol, 4), round(sharpe, 4)


def compute_time_series(weights: np.ndarray, returns_df: pd.DataFrame, tx_cost_bps: int):
    """Return equity curve, drawdown series, monthly returns and derived metrics."""
    port_returns = returns_df.values @ weights

    # Apply tx cost as a one-time drag on the first period
    equal = np.ones(len(weights)) / len(weights)
    cost = apply_tx_cost(equal, weights, tx_cost_bps)
    port_returns = port_returns.copy()
    if len(port_returns) > 0:
        port_returns[0] -= cost

    # Equity curve rebased to 100
    equity = 100.0 * np.cumprod(1 + port_returns)

    # Drawdown
    peak = np.maximum.accumulate(equity)
    dd_series = equity / peak - 1
    max_dd = float(dd_series.min())

    # Annualised return and Calmar
    n = len(equity)
    total_ret = equity[-1] / 100.0 if n > 0 else 1.0
    ann_return = float(total_ret ** (252.0 / n) - 1) if n >= 2 else 0.0
    calmar = round(ann_return / abs(max_dd), 4) if max_dd < 0 else 0.0

    # Turnover vs equal weight (one-way)
    turnover = round(float(np.sum(np.abs(weights - equal))), 4)

    dates = [str(d)[:10] for d in returns_df.index]
    equity_curve = [{"date": d, "value": round(float(v), 4)} for d, v in zip(dates, equity)]
    drawdown_out = [{"date": d, "drawdown": round(float(v), 6)} for d, v in zip(dates, dd_series)]

    # Monthly returns — group by year/month to avoid resampling freq string issues
    port_series = pd.Series(port_returns, index=returns_df.index)
    monthly_grouped = port_series.groupby([port_series.index.year, port_series.index.month]).apply(
        lambda x: float((1 + x).prod() - 1)
    )
    monthly_returns = [
        {"year": int(ym[0]), "month": int(ym[1]), "value": round(float(v), 6)}
        for ym, v in monthly_grouped.items()
    ]

    return equity_curve, drawdown_out, monthly_returns, round(max_dd, 4), calmar, turnover

File: rnef-backtest/src/lib/engine/metrics.ts
import { dot, colMeans, covMatrix, vecMatVec, cumProd, maxAccumulate } from './matrix'

function applyTxCost(before: number[], after: number[], bps: number): number {
  let turnover = 0
  for (let i = 0; i < before.length; i++) turnover += Math.abs(after[i] - before[i])
  return turnover * (bps / 10_000)
}

export function computeMetrics(
  weights: number[],
  returns: number[][],
  txCostBps: number
): { annReturn: number; annVol: number; sharpe: number } {
  const meanRet = colMeans(returns)
  const cov = covMatrix(returns)
  const n = weights.length

  const portReturn = dot(weights, meanRet) * 252
  const portVol = Math.sqrt(Math.max(vecMatVec(weights, cov), 0) * 252)

  // transaction cost vs equal weight baseline (annualized)
  const equal = new Array(n).fill(1 / n)
  const cost = applyTxCost(equal, weights, txCostBps) * 252
  const netReturn = portReturn - cost

  const sharpe = portVol > 0 ? netReturn / portVol : 0

  return {
    annReturn: Math.round(netReturn * 10000) / 10000,
    annVol: Math.round(portVol * 10000) / 10000,
    sharpe: Math.round(sharpe * 10000) / 10000,
  }
}

export function computeTimeSeries(
  weights: number[],
  returns: number[][],
  dates: string[],
  txCostBps: number
): {
  equityCurve: Array<{ date: string; value: number }>
  drawdown: Array<{ date: string; drawdown: number }>
  monthlyReturns: Array<{ year: number; month: number; value: number }>
  maxDD: number
  calmar: number
  turnover: number
} {
  const n = weights.length
  const nPeriods = returns.length

  // daily portfolio returns = returns_matrix @ weights
  const portReturns = returns.map(row => dot(row, weights))

  // apply one-time transaction cost on first day
  const equal = new Array(n).fill(1 / n)
  const cost = applyTxCost(equal, weights, txCostBps)
  portReturns[0] -= cost

  // equity curve rebased to 100
  const growthFactors = portReturns.map(r => 1 + r)
  const cumulative = cumProd(growthFactors)
  const equity = cumulative.map(v => 100 * v)

  // drawdown series
  const peak = maxAccumulate(equity)
  const ddSeries = equity.map((e, i) => e / peak[i] - 1)
  const maxDD = Math.min(...ddSeries)

  // annualized return from equity curve
  const totalRet = equity[equity.length - 1] / 100
  const annReturn = Math.pow(totalRet, 252 / nPeriods) - 1
  const calmar = maxDD < 0 ? annReturn / Math.abs(maxDD) : 0

  // turnover vs equal weight
  let turnover = 0
  for (let i = 0; i < n; i++) turnover += Math.abs(weights[i] - 1 / n)

  // format equity curve and drawdown
  const equityCurve = dates.map((d, i) => ({
    date: d,
    value: Math.round(equity[i] * 10000) / 10000,
  }))
  const drawdown = dates.map((d, i) => ({
    date: d,
    drawdown: Math.round(ddSeries[i] * 1000000) / 1000000,
  }))

  // monthly returns: group by year-month, compound within group
  const monthMap = new Map<string, number[]>()
  for (let i = 0; i < nPeriods; i++) {
    const d = new Date(dates[i])
    const key = `${d.getFullYear()}-${d.getMonth() + 1}`
    if (!monthMap.has(key)) monthMap.set(key, [])
    monthMap.get(key)!.push(portReturns[i])
  }

  const monthlyReturns: Array<{ year: number; month: number; value: number }> = []
  for (const [key, rets] of monthMap) {
    const [y, m] = key.split('-').map(Number)
    // compound: product of (1 + r) - 1
    const compounded = rets.reduce((acc, r) => acc * (1 + r), 1) - 1
    monthlyReturns.push({
      year: y,
      month: m,
      value: Math.round(compounded * 10000) / 10000,
    })
  }

  return {
    equityCurve,
    drawdown,
    monthlyReturns,
    maxDD: Math.round(maxDD * 10000) / 10000,
    calmar: Math.round(calmar * 10000) / 10000,
    turnover: Math.round(turnover * 10000) / 10000,
  }
}
